apps/04TowersOfHanoi.js

```diff
@@ -7,36 +7,69 @@ var rl = readline.createInterface({
   output: process.stdout
 });
 
+//object containing each stack
 var stacks = {
   a: [4, 3, 2, 1],
   b: [],
   c: []
 };
 
+//print the stacks to the console
 function printStacks() {
   console.log("a: " + stacks.a);
   console.log("b: " + stacks.b);
   console.log("c: " + stacks.c);
 }
 
-function movePiece() {
-  // Your code here
-
+function movePiece(startStack, endStack) {
+  //stores the last item in arrays 'a' 'b' or 'c'
+  var store = stacks[startStack].pop();
+  //pushes the stored item to 'a' 'b' or 'c'
+  stacks[endStack].push(store);
 }
 
-function isLegal() {
-  // Your code here
+//makes sure the only valid user inputs are 'a' 'b' or 'c'
+function validInput(startStack, endStack) {
+  if ((startStack === "a" || startStack === "b" || startStack === "c") && (endStack === "a" || endStack === "b" || endStack === "c")) {
+    return true;
+  } else {
+    return false;
+  };
+}
 
+function isLegal(startStack, endStack) {
+  //stores the length of 'a' 'b' or 'c'
+  var startStackLength = stacks[startStack].length;
+  var endStackLength = stacks[endStack].length;
+  //checks to make sure startStack is smaller than endStack and run movePiece function if so
+  if (stacks[endStack].length === 0 || stacks[startStack][startStackLength - 1] < stacks[endStack][endStackLength - 1]) {
+    movePiece(startStack, endStack);
+    return true;
+  } else {
+    console.log("Invalid move. Try again.");
+    return false;
+  };
 }
 
 function checkForWin() {
-  // Your code here
-
+  //checks lengths of 'b' or 'c' to see if win condition is met
+  if (stacks["b"].length === 4 || stacks["c"].length === 4) {
+    console.log("You won!");
+    return true;
+  } else {
+    return false;
+  };
 }
 
+//main fuction
 function towersOfHanoi(startStack, endStack) {
-  // Your code here
-
+  if (validInput(startStack, endStack) === true) {
+    isLegal(startStack, endStack);
+    checkForWin();
+  } else if (validInput(startStack, endStack) === false) {
```

Comment:
If `validInput(startStack, endStack)` doesn't return true, the only other option is a false value.

Thus, an `else` instead of an `else if` statement would be appropriate here.